### DISCUSSION TITLE
Correct usage of useQuery() in React app login form

### DISCUSSION THREAD

**Discussion by mateuszwojt** [urql-graphql/urql#2276] [Q&A] [answered]
I want to query the users' email and password and get the authorization tokens in return when the user submits the login form in my React app.

Here's the query:

```
  const [loginData, login] = useQuery({
    query: 
      `query ($email: String!, $password: String!) {
        login(params: {email: $email, password: $password}) {
          jwt, refreshToken
       }
    }`,
    variables: { email, password },
    pause: true,
  });
```

I'm using the pause: true argument to make sure it isn't triggered until the form is sent. In my handler function, I'm just firing the query (login) like a regular function:

```
  const submitHandler = (event) => {
    try {
      event.preventDefault();

      sessionStorage.clear();
      login(); // <-- here's the problem

      console.log(loginData);

      if (loginData?.data.login != null) {
        ...
      }

      ...
    } catch (error) {
      ...
    }
  };
```

This sometimes returns an error `TypeError: Cannot read properties of undefined (reading 'login')`, and sometimes it just lets me log in to the app. I'm not sure if I'm using this call correctly here. It returns an empty `loginData` object when the `TypeError` occurs.

**Comment by JoviDeCroock** [comment, marked as answer]
Let's start with the start, a login is better meant as a mutation this because a query will be cached and is intended to not have side-effects. When you are using a query like you are the info won't work out as you are opening a stream and hoping it delivers data in time.

The docs page to use it in a cb https://formidable.com/open-source/urql/docs/basics/react-preact/#using-the-mutation-result

**Reply by mateuszwojt** [reply]
I changed my query to mutation and everything works now flawlessly. Thanks, @JoviDeCroock !
